src/Week7/HW07.hs

```diff
@@ -0,0 +1,197 @@
+mapM f xs = sequence $ f <$> xs
+
+getElts :: [Int] -> Vector a -> Maybe [a]
+getElts xs vec = Week7.HW07.mapM (vec !?) xs
+
+-- Exercise 3 -----------------------------------------
+
+type Rnd a = Rand StdGen a
+
+randomElt :: Vector a -> Rnd (Maybe a)
+randomElt vec =
+    do
+        ind <- getRandomR (0, V.length vec -1)
+        return $ vec !? ind
+
+
+-- Exercise 4 -----------------------------------------
+
+randomVec :: Random a => Int -> Rnd (Vector a)
+randomVec n = V.fromList <$> replicateM n getRandom
+
+randomVecR :: Random a => Int -> (a, a) -> Rnd (Vector a)
+randomVecR n (lo,hi) = V.fromList <$> replicateM n (getRandomR (lo,hi))
+
+-- Exercise 5 -----------------------------------------
+
+swap :: (Int, Int) -> Vector a -> Vector a
+swap (i,j) vec = 
+    let jv = vec ! j in 
+    let iv = vec ! i in 
+    vec // [(i, jv), (j, iv)]
+
+shuffle :: Vector a -> Rnd (Vector a)
+shuffle vec = do
+    swaps <- Week7.HW07.mapM swapPartner [n-1,n-2 .. 1]
+    return $ foldr swap vec swaps
+        where
+            n = V.length vec
+            swapPartner i = (\j -> (i,j)) <$> getRandomR (0,n)
+
+
+-- Exercise 6 -----------------------------------------
+
+partitionAt :: Ord a => Vector a -> Int -> (Vector a, a, Vector a)
+partitionAt vec i = (less, pivotVal, more)
+    where
+        pivotVal = vec ! i
+        (less,more) = V.partition (< pivotVal) vec
+
+
+-- Exercise 7 -----------------------------------------
+
+-- Quicksort
+quicksort :: Ord a => [a] -> [a]
+quicksort [] = []
+quicksort (x:xs) = quicksort [ y | y <- xs, y < x ]
+                   <> (x : quicksort [ y | y <- xs, y >= x ])
+
+qsort :: Ord a => Vector a -> Vector a
+qsort = V.fromList . quicksort . V.toList
+
+
+-- Exercise 8 -----------------------------------------
+
+qsortR :: Ord a => Vector a -> Rnd (Vector a)
+qsortR vec
+    | V.null vec = return V.empty
+    | otherwise = details >>= helperSort
+        where
+            details = getRandomR (0, V.length vec -1) >>= return . partitionAt vec
+            helperSort (l,m,h) = liftM2 (<>) (qsortR l) (Control.Monad.Random.liftM (cons m) (qsortR h))
+
+
+
+
+-- Exercise 9 -----------------------------------------
+
+-- Selection
+select :: Ord a => Int -> Vector a -> Rnd (Maybe a)
+select idx vec 
+    | V.null vec = return Nothing 
+    | otherwise = chosen >>= helper
+        where 
+            chosen = getRandomR (0, V.length vec -1) >>= return . partitionAt vec
+            helper (l,m,h)
+                | idx < lLen = select idx l
+                | idx > lLen = select (idx - lLen - 1) h
+                | otherwise = return $ Just m 
+                where 
+                    lLen = V.length l
+
+-- Exercise 10 ----------------------------------------
+
+allCards :: Deck
+allCards = [Card label suit | suit <- suits, label <- labels ]
+
+newDeck :: Rnd Deck
+newDeck =  shuffle allCards
+
+-- Exercise 11 ----------------------------------------
+
+nextCard :: Deck -> Maybe (Card, Deck)
+nextCard d 
```

Comment:
[uncons](https://hackage.haskell.org/package/vector-0.12.3.0/docs/Data-Vector.html#v:uncons)?

Reply:
Oddly enough, it seems this function isn't being exported by Data.Vector, it doesn't show up even when I qualify the import with something like `Data.Vector.uncons` or `V.uncons`.  

Still trying to figure out why this is the case.

Reply:
It's because your [stack.yaml](https://github.com/dzhou343/CIS194/blob/master/stack.yaml#L21) is using resolver lts-17.11 which contains vector https://www.stackage.org/lts-17.11/package/vector-0.12.1.2 but uncons I linked there was only added in vector 0.12.2. 

It's minor detail, so don't worry about it.